packages/perspective-viewer-d3fc/src/ts/types.ts

```diff
@@ -0,0 +1,213 @@
+// ┃ ██████ ██████ ██████       █      █      █      █      █ █▄  ▀███ █       ┃
+// ┃ ▄▄▄▄▄█ █▄▄▄▄▄ ▄▄▄▄▄█  ▀▀▀▀▀█▀▀▀▀▀ █ ▀▀▀▀▀█ ████████▌▐███ ███▄  ▀█ █ ▀▀▀▀▀ ┃
+// ┃ █▀▀▀▀▀ █▀▀▀▀▀ █▀██▀▀ ▄▄▄▄▄ █ ▄▄▄▄▄█ ▄▄▄▄▄█ ████████▌▐███ █████▄   █ ▄▄▄▄▄ ┃
+// ┃ █      ██████ █  ▀█▄       █ ██████      █      ███▌▐███ ███████▄ █       ┃
+// ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
+// ┃ Copyright (c) 2017, the Perspective Authors.                              ┃
+// ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
+// ┃ This file is part of the Perspective library, distributed under the terms ┃
+// ┃ of the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0). ┃
+// ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+import { IPerspectiveViewerPlugin } from "@finos/perspective-viewer";
+import { DataRow, Type } from "@finos/perspective";
+
+export interface Element {
+    msMatchesSelector(selectors: string): boolean;
+}
+
+export interface Chart {
+    (container: any, settings: any): void;
+    plugin?: {
+        name: string;
+        category: string;
+        max_cells: number;
+        max_columns: number;
+        render_warning: boolean;
+        initial: {
+            type?: string;
+            count?: number;
+            names: string[];
+        };
+        selectMode?: string;
+    };
+}
+
+export type PadUnit = "percent" | "domain";
+export type Pad = [number, number];
+
+export type GetSetReturn<T, Arg, Func> = T extends undefined
+    ? Arg
+    : T extends Arg
+    ? Func
+    : never;
+
+export interface PaddingStrategy {
+    (extent: any): any;
+    pad: <T extends Pad | undefined = undefined>(
+        nextPad?: T
+    ) => GetSetReturn<T, Pad, PaddingStrategy>;
+    padUnit: <T extends PadUnit | undefined = undefined>(
+        nextPadUnit?: T
+    ) => GetSetReturn<T, PadUnit, PaddingStrategy>;
+}
+
+export type Axis = [number, number];
+
+export type GradientPair = [number, string];
+
+export type GradientKey = string; // full, negative, positive
+
+export type ColorStyles = {
+    scheme: string[];
+    gradient?: Record<GradientKey, GradientPair[]>; // Should this be optional?
+    interpolator: { [key: string]: (value: number) => string };
+    grid: { gridLineColor?: string };
+    opacity?: number;
+    series?: string;
+    [key: `series-${number}`]: string;
+};
+
+export type TextStyles = Record<string, string>; // Should this be optional?
+
+export type DataRowsWithKey = DataRow[] & {
+    key?: string;
+};
+
+export type MainValue = {
+    name: string;
+    type: Type;
+};
+
+// NOTE: Should these props be optional?
+export type TreemapValue = {
+    treemapLevel?: number;
+    treemapRoute?: any[]; // string[]?
+};
+
+// Are these a particular type of settings?
+// Question: Is there a better type def for this?
+export type Settings = {
```

Comment:
Good catch. What do you think the type of `settings.columns` should be? Thanks for pointing out `EXCLUDED_SETTINGS`.

Reply:
`Record<string, Record<string, unknown>>`

This could probably be narrowed so the inner Record is an object containing the specific types of plugin configurations. For example:

```typescript
type ColumnConfig = {
    symbol: Record<string, string>
}
type ColumnSettings = Record<string, ColumnConfig>
```

This particular configuration setting is undergoing changes atm so don't worry about fitting it too closely

Reply:
Awesome, I've updated the `settings.columns` type.